### ISSUE TITLE
Add "External Build Order"

### ISSUE THREAD

**Issue by SchrodingersGat** [inventree/InvenTree#1071]
Add a concept which (kind of) combines the ideas of a PurchaseOrder and a BuildOrder

- An external supplier provides a "process" which converts one part into another
- e.g. an anodization step turns "un-anodized widget" into "anodized widget"
- e.g. an external PCB assembler loads components onto a board
- Needs to take items out of stock and add new items into stock
- If items are tracked, need to make use of the `install_into` function
- Also retain some of the concepts of a PurchaseOrder

**Comment by eeintech** [comment]
I'm not sure I see the purpose of tracking external builds. It makes sense to track items and processes inside a same entity/company, however outside jobs just turn in as "finished goods" therefore purchase/external build order leads to parts and stocks. Maybe I just don't see what kind of application would this be used for.

**Comment by SchrodingersGat** [comment]
Agreed it's a weird case. This is at the request of a friend of mine who has a very particular use case. I'm not 100% sure what the actual implementation looks like yet, this issue is (for now) mostly just a placeholder until i work out if it necessarily makes sense.

**Comment by jlaw** [comment]
I think I am looking for this use case. I want to use InvenTree to track PCB manufacturing:
1. Order 15 bare boards from PCB fab (supplier+manufacturer).
2. Receive 15.
3. Order 10 assembled boards from PCB assembler (supplier+manufacturer). They require 10 bare boards from me, so I need to supply 10 to them from my 15 stock.
4. Receive 10.
5. Sell 5 to customer.
6. Customer wants 10 more.
7. Place order with PCB assembler for 5 more, taking the remaining 5 bare boards.
8. Receive 5.
9. Take 5 new + 5 existing in inventory to sell to customer.

**Comment by eeintech** [comment]
@jlaw Instead of "selling" your PCB + BOM to your PCB assembler, why not creating build orders for that process?
Granted, your PCB and other parts will still show as "in-stock" in your InvenTree but once they are allocated to the build they become essentially unusable for anything else. So that would be one way I would do it.

**Comment by jlaw** [comment]
Yes, that is one way, but I would essentially have to double enter the order: once as a BO and the second as a PO, which I think this issue was suggesting to combine.

**Comment by eeintech** [comment]
@jlaw Ha right! I kinda lost track of the initial idea :sweat_smile:

**Comment by SchrodingersGat** [comment]
@jlaw this is a perfect example of this use-case - an external supplier that you need to also supply parts to. It's a mix of a "purchase order" and a "build order" in a way.

InvenTree does not really support this yet, but I think it should! I just don't have a picture in my head yet of how to implement it.

Maybe we could add the concept of assigning internal stock against a purchase order? So the stock is consumed by the order like it would be for a build order?

**Comment by eeintech** [comment]
@jlaw I played with purchase orders, sales orders and build orders with the latest InvenTree version, for the same scope of PCB assembly manufacturing and distribution.

Here are my thoughts for your workflow:
> * Order 15 bare boards from PCB fab (supplier+manufacturer).

Associate supplier to PCB part and create PO for PCB.

> * Receive 15.

Complete PO

> * Order 10 assembled boards from PCB assembler (supplier+manufacturer). They require 10 bare boards from me, so I need to supply 10 to them from my 15 stock.

Create build order for PCB assembly, allocate solely the PCB parts

> * Receive 10.

Complete build order

> * Sell 5 to customer.

Create sales order

> * Customer wants 10 more.

Create new sales order, create new build order from sales order

> * Place order with PCB assembler for 5 more, taking the remaining 5 bare boards.

Allocate PCB part only in new build order

> * Receive 5.

Complete build order

> * Take 5 new + 5 existing in inventory to sell to customer.

Complete new sales order.

I don't **think** InvenTree is currently missing any logic other than "annotating" build orders as being fulfilled externally.
We could also associate a "price" tag to the "external" build order, but the rest of the flow should work.
What do you think?

For our company I am going to propose a similar flow with existing features.

**Comment by jlaw** [comment]
@eeintech Thanks for breakdown.

I think to solve this would require allowing a build to be add-able to a PO like you said. Somehow make a build be a 'part'.

I'm unsure if the location of the build order and location of the parts to build the order needs to be at the vendor/suppliers location. If we don't need to check or track this, then we may not be able to account for shipping and tracking delivery times., which may or may not be a feature now or in the future.

**Comment by miggland** [comment]
I saw this issue (again) because of #4044.

Just wanted to express my interest, ie add that we have a similar process to @jlaw - the background being that we order parts, designed by us, from various manufacturers. We then have external companies combine these parts into others - ie perform a Build Order. We need to supply them from our stock. A Build Order is great to keep track of which parts are required, just "selling" the stock to the external party wouldn't really be a good solution.

I use the available functionality to cover this - ie a Build Order, and have a "stock location" which represents the external manufacturer.

What's missing is having the data ready for actually sending the order to the external manufacturer.
The connection of the BO to an external company, as @eeintech pointed out, would allow creating a report/webhook/... whatever to actually send the order to the right place.

What's not covered is the fact that BOs have a different numbering system - creating an order from the BO would "mess up" the list of POs by having a different reference system. This is not a big issue, and I could probably fix this by changing our process to take this into account. I see no easy solution to this problem in InvenTree either.

**Comment by JambonLeVrai** [comment]
Bumping this to show my interest toward this functionality. 
Just like most of the people here, I have mechanical parts that come unpainted, and we outsource the painting operation. Currently what we do to work around this is that we have a "part painting" part for each mechanical part we paint, and we make a purchase order for these paintings. Then we combine the unpainted part with the painting through a build order to get the painted part. It is working and keeps the prices up-to-date but it is quite a bit tedious.
I also have a situation which I don't believe has been talked about here, with manufacturing assembled PCBs. We need to supply only some components, and they will take care of the rest. If we want to track the stock of our components, and keep the price of the whole PCB, it is also getting tedious. What I have thought of is also making two different parts, one part is the "pcb manufacturing and assembly" purchased from the manufacturer, and another part which is the "complete pcb", which would be an assembly in Inventree with both the "pcb manufacturing and assembly" and the other components we need to send. This also mean that it is getting more complicated if the components we need to send to the manufacturer vary from one order to another (with the component shortages, sometimes the manufacturer have the chips, sometimes they do not and we have to send some), which would mean multiple BoMs (and therefore probably creating different "variants" of the same actual part with different BoMs).
The issue with making a system with only a purchase order that allows to subtract some specified stock would be with the actual pricing information for the resulting part, since the total cost of the final part would be the cost of the outsourcing operation + the initial cost of the part.
I believe that the simplest way would be to add an "Outsourcing BoM" to the final part. Then the part could be bought using "Outsourcing orders", which would need stock to be allocated just like a Build order, but be placed like a Purchase order, along with a price / part. That way we can keep track of the real price of the part.

**Comment by matmair** [comment]
This is a pretty involved idea but I am sure the team would be willing to provide some pointers and possible data flows if someone steps up to to the coding.

**Comment by SchrodingersGat** [comment]
@JambonLeVrai thanks for adding your thoughts! 

> Currently what we do to work around this is that we have a "part painting" part for each mechanical part we paint, and we make a purchase order for these paintings. Then we combine the unpainted part with the painting through a build order to get the painted part. 

I can see the "appeal" of your approach, but also that it results in multiple InvenTree processes where perhaps a single step would be simpler. 

> manufacturing assembled PCBs. We need to supply only some components, and they will take care of the rest.

I do this a *lot* and it does really result in complex pricing data which is hard to integrate. An "outsourced BOM" or "external build order" which can handle all of the different requirements listed in this thread sounds pretty complex.

**Comment by Adamantan** [comment]
Just to chime in with the ones who spoken before on this thread.
We have the exact same use case. Order bare PCB-s from fab in china, ship bare PCB-s to external EMS in Europe, receive finished PCB-s. Just as mentioned before we are also having trouble tracking the pricing of the assembled boards without adding an extra in house "build order" combining the "assembly service" and the "bare board" parts.

Looking forward to see a feature for this whenever it makes it onto the roadmap. :)

**Comment by SchrodingersGat** [comment]
@Adamantan any suggestions as to how this might be cleanly implemented would be welcome :)

**Comment by Adamantan** [comment]
Here you have two ways to go I think, without knowing much about how the code actually functions:
The long way: You can rely on BOM lists and Build orders. You could have an "external build order" which would substitute any or all components in the output devices BOM list with a single "cost" of the external manufacturer. The parts which were not substituted are still consumed from stock, so you can have the effect that you send something away to a different company to be modified. This leaves a lot of options open for setting up your externally manufactured parts, but it is also pretty complex to keep track of.

The simpler way: My suggestion would be to use the PO feature with modifications. When you are buying from an external assembly house or the sorts you will need to hand them a purchase order anyways. Making an "External Build Order" PO would need a feature which would allows it to consume stock. You could have a "tick box" saying "external build order" when you make a PO. This would add a section like "extra line items" to the PO, where you can browse your stock and assign stock items to "send away" for modification. At issuing the PO, the selected stock would be consumed. The disadvantages are that there are no "safeguards" for assigning the wrong items to be sent away.

**Comment by SchrodingersGat** [comment]
Hey everyone, finally have a proof of concept in the works for this - see https://github.com/inventree/InvenTree/pull/9312

Any discussions / feedback on this implementation should be directed over to the linked PR
